src/main/java/net/minecraftforge/event/world/BlockEvent.java

```diff
@@ -282,7 +283,43 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
             super(world, pos, state);
         }
     }
-    
+
+    /**
+     * Fired when a liquid places a block. Use {@link #getNewState()} to change the result of
+     * a cobblestone generator or add variants of obsidian. Alternatively, you  could execute
+     * arbitrary code when lava sets blocks on fire, even preventing it.
+     *
+     * It should be noted that {@link #getState()} will return the block that was originally going to be placed.
+     */
+    @Cancelable
+    public static class FluidPlaceBlockEvent extends BlockEvent
+    {
+        private final BlockPos liquidPos;
+        private IBlockState newState;
+
+        public FluidPlaceBlockEvent(World world, BlockPos pos, BlockPos liquidPos, IBlockState state)
+        {
+            super(world, pos, state);
+            this.liquidPos = liquidPos;
+            this.newState = state;
+        }
+
+        public BlockPos getLiquidPos()
+        {
+            return liquidPos;
+        }
+
+        public IBlockState getNewState()
+        {
+            if (this.isCanceled()) return this.getWorld().getBlockState(this.getPos());
```

Comment:
this usage does not seem very fine

Reply:
Could you elaborate for me?

Reply:
You should store the original state in a field and add a getter for it.
For this line, I'm not sure it makes sense to return the original state when the event is canceled. Usually canceled just means "nobody after me gets the event", having it return the original state here is a bit unexpected I think.

Reply:
Well, alright.